Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitImmersiveMode.cpp

```diff
@@ -0,0 +1,152 @@
+        {
+            // Ignore server certificate errors.
+#if ENABLE(2022_GLIB_API)
+            m_originalPolicy = webkit_network_session_get_tls_errors_policy(m_test->m_networkSession.get());
+            webkit_network_session_set_tls_errors_policy(m_test->m_networkSession.get(), policy);
+#else
+            m_manager = webkit_web_context_get_website_data_manager(m_test->m_webContext.get());
+            m_originalPolicy = webkit_website_data_manager_get_tls_errors_policy(m_manager);
+            webkit_website_data_manager_set_tls_errors_policy(m_manager, policy);
+#endif
+        }
+        ~NetworkPolicyGuard()
+        {
+#if ENABLE(2022_GLIB_API)
+            webkit_network_session_set_tls_errors_policy(m_test->m_networkSession.get(), m_originalPolicy);
+#else
+            webkit_website_data_manager_set_tls_errors_policy(m_manager, m_originalPolicy);
+#endif
+        }
+        ImmersiveModeTest* m_test;
+        WebKitTLSErrorsPolicy m_originalPolicy;
+#if !ENABLE(2022_GLIB_API)
+        WebKitWebsiteDataManager* m_manager;
+#endif
+    };
+};
+
+#if USE(SOUP2)
+static void serverCallback(SoupServer* server, SoupMessage* message, const char* path, GHashTable*, SoupClientContext*, gpointer)
+#else
+static void serverCallback(SoupServer* server, SoupServerMessage* message, const char* path, GHashTable*, gpointer)
+#endif
+{
+    if (soup_server_message_get_method(message) != SOUP_METHOD_GET) {
+        soup_server_message_set_status(message, SOUP_STATUS_NOT_IMPLEMENTED, nullptr);
+        return;
+    }
+
+    auto* responseBody = soup_server_message_get_response_body(message);
+
+    if (g_str_equal(path, "/xr-session/")) {
+        soup_server_message_set_status(message, SOUP_STATUS_OK, nullptr);
+        soup_message_body_append(responseBody, SOUP_MEMORY_STATIC, indexHTML, strlen(indexHTML));
+        soup_message_body_complete(responseBody);
+    } else {
+        g_assert(false);
+        soup_server_message_set_status(message, SOUP_STATUS_NOT_FOUND, nullptr);
+    }
+}
+
+static void testWebKitImmersiveModeLeaveImmersiveMode(ImmersiveModeTest* test, gconstpointer)
+{
+    auto httpsServer = makeUnique<WebKitTestServer>(WebKitTestServer::ServerHTTPS);
+    httpsServer->run(serverCallback);
+
+    ImmersiveModeTest::NetworkPolicyGuard guard(test, WEBKIT_TLS_ERRORS_POLICY_IGNORE);
+
+    g_assert_false(test->isImmersiveModeEnabled());
+
+    test->loadURI(httpsServer->getURIForPath("/xr-session/").data());
+    test->waitUntilLoadFinished();
+
+    GUniqueOutPtr<GError> error;
+    test->runJavaScriptAndWaitUntilFinished("document.getElementById('enterXR').click();", &error.outPtr(), nullptr);
+    g_assert_no_error(error.get());
+#if ENABLE(WEBXR)
+    ConsoleMessageTest::ConsoleMessage referenceMessage = { ConsoleMessageTest::MessageSource::ConsoleAPI, ConsoleMessageTest::MessageLevel::Error, "XR session failed to start: SecurityError: The operation is insecure.", 1, httpsServer->getURIForPath("/xr-session/").data() };
```

Comment:
I don't understand the test, this is an api test, we want to check that our property changes when a session is created and that we can leave the session from the api. If we can't test that, it doesn't make sense to add an api test. To check that requestSession gives a security error we can use a layout test instead.

Reply:
We can't enter a session programmatically, because click doesn't focus the document.

If there is a way to focus the document, then we can enter the session, and then we can also exit it successfully

Reply:
I commented out the focus requirement (because it's no longer present in the upstream version of the spec), and then the test can be more interesting.

I am not sure we can relax the webxr condition though, from reading the spec it looks like not asking for document focus is covered by an earlier check. 

I'll share my findings in the topic on Monday

Reply:
However, this test is an answer to @mcatanzaro ask for a simple test. If this doesn't cut it, I'm ok removing the test but still submitting the API changes, at least until we have a mock XR coordinator where a test could return.

Reply:
I got this working by combining

```
        showInWindow();
        runJavaScriptAndWaitUntilFinished("document.getElementById('enterXR').focus()", nullptr);
        runJavaScriptAndWaitUntilFinished("document.getElementById('enterXR').click();", nullptr);
```

With that, I can enter an XR session **if Monado is running**. But if it isn't running, then the session simply doesn't start, so we can't test much -- only that the exception is thrown in JS.